### PR TITLE
Hive: Pre-check namespace presence for listTables

### PR DESCRIPTION
**Pull request by haizhou-zhao**
## Change
Add a Pre-check of namespace presence for listTables in HiveCatalog

## Intention
When invoking listTables on the current HiveCatalog for a database that does not exist, it will return an empty list; however, it seems the intention is to throw an `NoSuchNamespaceException` on such cases.

Currently, there is a catch-block intended to capture `UnknownDBException` thrown by Hive Metastore:
https://github.com/apache/iceberg/blob/7e2920a/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java#L163

However, the problem with that is Hive Metastore in many cases won't throw an `UnknownDBException` when listing tables in a database that does not exist: internally to Hive Metastore, databases and tables are tracked by separate RDBMS tables - a query against the "tables" table for an absent DB yields empty list, but it won't be aware of the absence of the DB.

### CODE REVIEW

**Comment by pvary** [comment]
@haizhou-zhao: which versions of HMS return empty list, and which version return `UnknownDBException`?
I'm specifically interested in the supported HMS versions (2.

**Comment by haizhou-zhao** [comment]
@pvary 

I know hive 2.3.9 and 3.1.2 returns empty list, which is close but a bit different from what you are asking for. Let me know if you are looking for 2.3.10 and 3.1.3 exactly.

I personally have never used hive 4.0.0 before, and I need to do some extra test to be able to tell you.

**Comment by pvary** [comment]
Maybe @deniskuzZ could help with info about Hive 4.0.0

**Comment by haizhou-zhao** [comment]
@pvary I downloaded Hive 4.0.0 binary and did a quick test. I believe listing tables on a DB that does not existing in 4.0.0 will also return an empty list, not throwing `UnknownDBException`.

**Comment by deniskuzZ** [comment]
Hive is throwing an exception when trying to list tables in non-existing catalog:
````
show tables in UNKNOWN;
INFO  : Compiling command(queryId=hive_20240807072922_373feca6-fccd-4d18-9379-c8c02106ce4c): show tables in UNKNOWN
ERROR : FAILED: SemanticException [Error 10072]: Database does not exist: UNKNOWN
````

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java [hunk 1]
```diff
@@ -351,6 +352,11 @@ public void testListTables() throws TException, IOException {
 
     assertThat(catalog.tableExists(TABLE_IDENTIFIER)).isTrue();
     HIVE_METASTORE_EXTENSION.metastoreClient().dropTable(DB_NAME, hiveTableName);
+
+    // test listing tables in a db that doesn't exist
+    assertThatThrownBy(() -> catalog.listTables(Namespace.of("db_does_not_exist")))
```
Comment (pvary) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java:357]: This is a different "feature", so it should be a different test method

**Comment by pvary** [comment]
So basically we don't get the expected results from the listing of the tables, so we issue another HMS call (and accept the performance degradation), so we can return the correct answer ourselves.

Please separate out the test for a different test method, then I think we are ok with the fix

**Pull request closed without merging**